smarts/core/sumo_road_network.py

```diff
@@ -269,6 +272,58 @@ def _snap_internal_edges(self, lane_to_poly, snap_threshold=2):
                 lane_shape = Polygon(snap(lane_shape, outgoing_shape, snap_threshold))
                 lane_to_poly[lane_id] = lane_shape
 
+    def _snap_internal_holes(self, lane_to_poly, snap_threshold=2):
+        for lane_id in lane_to_poly:
+            lane = self.lane_by_id(lane_id)
+
+            # Only do snapping for internal edge lane holes
+            if not lane.getEdge().isSpecial():
+                continue
+            lane_shape = lane_to_poly[lane_id]
+            for x, y in lane_shape.exterior.coords:
+                nl = self.nearest_lane(
+                    (x, y),
+                    max(10, 2 * self._default_lane_width),
+                    include_junctions=False,
+                )
+                if nl:
+                    nl_shape = lane_to_poly.get(nl.getID())
+                    lane_shape = Polygon(snap(lane_shape, nl_shape, snap_threshold))
+                    lane_to_poly[lane_id] = lane_shape
+
+    def _snap_external_holes(self, lane_to_poly, snap_threshold=2):
+        for lane_id in lane_to_poly:
+            lane = self.lane_by_id(lane_id)
+
+            # Only do snapping for external edge lane holes
+            if lane.getEdge().isSpecial():
+                continue
+
+            incoming = self.lane_by_id(lane_id).getIncoming()
+            if incoming:
+                if incoming[0].getEdge().isSpecial():
+                    continue
+
+            outgoing = self.lane_by_id(lane_id).getOutgoing()
+            if outgoing:
+                outgoing_lane = outgoing[0].getToLane()
+                if outgoing_lane.getEdge().isSpecial():
+                    continue
+
+            lane_shape = lane_to_poly[lane_id]
+            for x, y in lane_shape.exterior.coords:
+                nl = self.nearest_lane(
+                    (x, y),
+                    max(10, 2 * self._default_lane_width),
+                    include_junctions=False,
+                )
+                if nl:
+                    if nl.getEdge().isSpecial():
+                        continue
+                    nl_shape = lane_to_poly.get(nl.getID())
+                    lane_shape = Polygon(snap(lane_shape, nl_shape, snap_threshold))
+                    lane_to_poly[lane_id] = lane_shape
```

Comment:
Same two comments as above:
- is it ok if `nl_shape` is `None` in the call to `snap()`?
- not good to update dict while iterating over it

Reply:
Yup, will make these changes.